web/src/search/results/SearchResults.tsx

```diff
@@ -216,6 +226,46 @@ export class SearchResults extends React.Component<SearchResultsProps, SearchRes
                 )
         )
 
+        this.subscriptions.add(
+            this.componentUpdates
+                .pipe(
+                    startWith(this.props),
+                    distinctUntilChanged((a, b) => isEqual(a.location, b.location))
+                )
+                .subscribe(props => {
+                    const searchParams = new URLSearchParams(props.location.search)
+                    const versionFromURL = searchParams.get('c')
+
+                    if (
+                        !searchParams.has('from-context-toggle') &&
+                        props.availableVersionContexts &&
+                        (versionFromURL
+                            ? versionFromURL !== localStorage.getItem(LAST_VERSION_CONTEXT_KEY)
+                            : localStorage.getItem(LAST_VERSION_CONTEXT_KEY) !== null)
+                    ) {
+                        this.setState({ showVersionContextWarning: true })
+                    }
+
+                    if (searchParams.has('from-context-toggle')) {
+                        searchParams.delete('from-context-toggle')
+                        this.props.history.replace({
+                            search: searchParams.toString(),
+                            hash: this.props.history.location.hash,
+                        })
+                        this.setState({ showVersionContextWarning: false })
+                    }
+
+                    if (
+                        this.state.showVersionContextWarning &&
+                        (versionFromURL
+                            ? versionFromURL === localStorage.getItem(LAST_VERSION_CONTEXT_KEY)
+                            : localStorage.getItem(LAST_VERSION_CONTEXT_KEY) === null)
+                    ) {
+                        this.setState({ showVersionContextWarning: false })
+                    }
```

Comment:
I spent a bit of time wrapping my head around these 3 if statements - I think this is simpler yet essentially equivalent. Please double-check my logic though 🙂 The ternary on `versionFromURL` + `localStorage.getItem(LAST_VERSION_CONTEXT_KEY) === null` checks aren't necessary, since `URLSearchParams.get()` returns `null` when a key isn't present.

```suggestion
                    if (searchParams.has('from-context-toggle')) {
                        searchParams.delete('from-context-toggle')
                        this.props.history.replace({
                            search: searchParams.toString(),
                            hash: this.props.history.location.hash,
                        })
                        this.setState({ showVersionContextWarning: false })
                    } else (
                        this.setState({
                            showVersionContextWarning: props.availableVersionContexts && versionFromURL !== localStorage.getItem(LAST_VERSION_CONTEXT_KEY)
                        })
                    }
```

Reply:
Thanks a lot, this works and is way nicer! 

Implemented this separately since there were acouple typos